Faxing of unusual objects

Jacquerel opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

The other day I was playing as a clown, happily photocopying my ass, only to come to the terrible realisation that fax machines do not allow you to fax photos of your ass to the heads of staff. This PR fixes that.
I threw in tcg cards and individual sheets of carboard too, because those are _basically_ paper.

Additionally, I added a wires panel to fax machines.
Alongside the usual shock wire, the options here are:
- Hide the fax machine from the network (it cannot receive faxes).
- Overcharge input servos (allows some additional items to be faxed).
- Overcharge output servos (received items are thrown in a random direction instead of placed daintily).

The overcharged input tray allows some additional fun items, largely dictated by things which I figured "are flat" and might be either useful or funny to fax to someone.
- Individual sheets of most materials, if their icon is just a rectangle (so yes iron, no gold bars).
- ID Cards (and the emag/doorjack).
- Flatbread.
- Individual slices of pizza.
- Throwing stars (combine with hacking the output tray for the world's least accurate or efficient assassination method).

Faxing unusual materials to a machine has a 20% chance to jam its tray, at which point it cannot receieve any faxes until it has been cleaned out.

https://user-images.githubusercontent.com/7483112/188020403-8b92c263-b5f9-4879-8dcb-102cd27de272.mp4

In this example video I didn't know that it wasn't capturing the tgui panels so uh... just imagine those are there. They look the same anyway.

In case you are curious my reasoning for making the allowed types lists static on fax machine instead of just defines is in case an admin wanted to add additional faxable items mid round for some fax-related gimmick.
If you can think of any flat and vaguely paper-like item that I forgot about which would be good for fax machines please let me know.

While I was there I added a bunch of screentips too, it turns out the fax machine interacts with a lot of items.

## Why It's Good For The Game

More interactions with the fax machine allow it to become a more versatile part of the sandbox.
The ability to fax photos really just felt like an obvious omission, I have seen several different people other than myself attempt to do this and then be disappointed when it did not work.
It would be funny for a chef to spend a round barricaded in the kitchen and only interacting with the outside world by faxing slices of pizza to their table.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: You can now send photos or trading cards through the fax machine.
add: You can now hack the wires of a fax machine to have it throw items harder or accept additional flat input items such as material sheets, ID cards, and pizza slices.
qol: Screentips should tell you what using a particular item on a fax machine will do.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Jacquerel committed:
Fax machines have a wire panel.
Fax machines can be hacked to allow unusual input, throw their contents on receipt, and be invisible to the network.
Fax machines will accept some additional paper-like items by default.
Fax icons moved to their own file.

Wallemations commented:
Does the machine not have an open wires overlay sprite?

Jacquerel commented:
> Does the machine not have an open wires overlay sprite?

![image](https://user-images.githubusercontent.com/7483112/188029781-dd11ca34-253c-4c03-b38c-f14441aff637.png)
Might be hard to see in the video due to compression to get it into the file size limit.

Baskett55 commented:
Money would be a cool thing to fax too. Bribe the captain or centcom official for something extra.

twilightwanderer commented:
In my opinion, you are very overcomplicating the mechanics of the fax machine and making it into something it shouldn't be doing by turning it into a teleporter for things.

rasepretep commented:
Whatever weirdness this PR brings is directly caused by the original PR making it teleport items instead of working like a fax. Unless that's fixed, we might as well have fun and send pizza slices. It's rarely gonna be used anyway.

twilightwanderer commented:
Okay, you have convinced me. Then I'll wait until they include this PR in the build to adjust my forthcoming update for the admin panel. To be honest, when I was developing this fax, I didn't expect such interest from other developers.

Jacquerel commented:
@twilightwanderer @MrMelbert If you think I am overcomplicating the interactions with this machine I am happy to make changes. I had different thought processes here based on what aspect we're talking about.
Photographs are really a big one which I have seen people try to do already and then be surprised that it doesn't work, a normal fax machine would be able to do this so this seems like an eminently reasonable addition to me.

Several of the other items (money, ID cards, trading cards) would _not_ work in a real fax machine but you could fax an approximation of them. In a similar vein to photocopying secret documents or your own butt I don't think people would be that surprised at it working this way especially if they are aware that the fax machine is _not_ creating copies like it would in real life.

And then the more exotic items I will admit are just jokes based on items I think you could shove into the slot on a machine, but I _do_ think there is also a cultural understanding of the idea of putting food into computers (if we had individual slices of ham I definitely would have put them on the list) and people will get a kick out of trying it and then it actually working.

I am happy to remove any or all of the "exotic items" if push somes to shove and you don't think it would be a good addition. I would be happy for this PR to mostly just encompass faxing photos and adding a wire panel which flings paper ineffectually around the room (maybe I'd make the input override wire crumple incoming papers or fold them into aeroplanes instead), but I do think this would be a good addition even if only for extremely niche reasons.

And as my final justification, this: http://gunshowcomic.com/381

Jacquerel committed:
Unpunctuates screentips.
You can send cash now also.

Likteer commented:
@Jacquerel We don't have ham slices, but we do have salami (slice a sausage after cooking it)!

Jacquerel committed:
Merge remote-tracking branch 'tgstation/master' into pizza-fax
Adds logging to non-paper faxes.

Jacquerel committed:
Removes unnecessary "var".
Adds salami to food list.

Wallemations commented:
If this ends up not getting merged I still think making a PR for the throwing of paper & removing from the network would be good

Jacquerel commented:
> If this ends up not getting merged I still think making a PR for the throwing of paper & removing from the network would be good

Yes, if required I would just remove the joke additions and keep that.

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -12,34 +12,85 @@
+	if(seconds_electrified > MACHINE_NOT_ELECTRIFIED)
+		seconds_electrified -= delta_time
+
+/obj/machinery/fax/_try_interact(mob/user)
```
This should be on ``attack_hand()``

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -49,10 +100,24 @@
+/**
+ * Open and close the wire panel.
+ */
+/obj/machinery/fax/screwdriver_act(mob/living/user, obj/item/screwdriver)
+	if(..())
+		return TRUE
```
```suggestion
/obj/machinery/fax/screwdriver_act(mob/living/user, obj/item/screwdriver)
	. = ..()
```
parent does nothing

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -110,56 +231,155 @@
+		var/obj/item/paper/sent_paper = sent
+		log_paper("[usr] has sent a fax with the message \"[sent_paper.get_raw_text()]\" to [name]/[destination_id].")
+		return
+	if (istype(sent, /obj/item/throwing_star))
+		log_attack("[usr] has faxed [sent] to [name]/[destination_id].]")
+		return
+	log_game("[usr] has faxed [sent] to [name]/[destination_id].]")
```
logging in 2 different places for faxing seems too confusing for admins to look through

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -110,56 +231,155 @@
+	var/mutable_appearance/overlay = mutable_appearance(icon, overlay_state)
+	colour_material_overlay(item, overlay)
+	overlays += overlay
+	sleep(2 SECONDS)
+	icon_state = "fax"
+	overlays -= overlays
```
add a timer for end animate object travel, instead of sleep.

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -12,34 +12,85 @@
 	var/list/fax_history = list()
 
+	/// List of types which should always be allowed to be faxed
+	var/static/list/allowed_types = list(/obj/item/paper, /obj/item/photo, /obj/item/tcgcard, /obj/item/stack/sheet/cardboard,)
```
```suggestion
	var/static/list/allowed_types = list(
		/obj/item/paper,
		/obj/item/photo,
		/obj/item/tcgcard,
		/obj/item/stack/sheet/cardboard,
	)
```

Lists should be below all the other vars

JohnFulpWillard reviewed (changes requested):
I disagree with sending materials through faxes, it will only cause problems if the material's sprites ever change, since you just have a whitelist for what materials are (currently) flat.

You added a ton of things to fax, some are really funny, but I don't see any upsides of being able to fax materials, and since it's more work than the rest (like pizza), I think there should have a very good reason for it, or we should not add it at all.

Jacquerel committed:
Removes material sheets.
All non-paper faxes are logged to game, paper ones are still logged to paper as they contain parsed text.
Removes sleeping.

Jacquerel committed:
Suggested formatting changes.
Machine should correctly reregister to process.

JohnFulpWillard commented on code/datums/wires/fax.dm:
```diff
@@ -0,0 +1,50 @@
+	..()
+
+/datum/wires/fax/interactable(mob/user)
+	if(!..())
+		return FALSE
```
```suggestion
	. = ..()
	if(!.)
		return FALSE
```

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -12,34 +12,92 @@
 	/// Necessary to hide syndicate faxes from the general list. Doesn't mean he's EMAGGED!
 	var/syndicate_network = FALSE
 	/// This is where the dispatch and reception history for each fax is stored.
 	var/list/fax_history = list()
```
put this below jammed since it's a list, please

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -12,34 +12,92 @@
+
+
+/obj/machinery/fax/on_set_is_operational(old_value)
+	if (old_value == FALSE)
```
```suggestion
	if (!old_value)
```

Jacquerel replied:
I decided to make it explicit `== FALSE` here because I feel like it just is easier to read that way, but I'll change it if it's required.

JohnFulpWillard replied:
That's fair I suppose, especially since ``old_value`` is a very uninformative var name.

JohnFulpWillard commented on code/datums/wires/fax.dm:
```diff
@@ -0,0 +1,50 @@
+/datum/wires/fax/New(atom/holder)
+	wires = list(WIRE_SHOCK, WIRE_SIGNAL, WIRE_THROW, WIRE_LOADCHECK,)
+	add_duds(1)
+	..()
```
```suggestion
	return ..()
```

JohnFulpWillard commented on code/modules/paperwork/fax.dm:
```diff
@@ -65,14 +136,68 @@
+/obj/machinery/fax/proc/can_load_item(obj/item/item)
+	if(!is_allowed_type(item))
+		return FALSE
+	if (!istype(item, /obj/item/stack))
+		return TRUE
+	var/obj/item/stack/stack_item = item
+	return stack_item.amount == 1
```
Is this still needed? Do we want to prevent space cash transfering to 1 at a time?

Jacquerel replied:
I think it both makes more sense and also is funnier to be limited to sending one bill at a time rather than an entire stack, but if you think it's too pedantic (or isn't worth the extra complication) I can cut it out.

JohnFulpWillard replied:
It's fine as is, was asking if you had a different opinion with materials gone.

JohnFulpWillard reviewed (commented):
last review I believe

Jacquerel committed:
Reorders properties.
Some refactoring of returns and calls to super.
